glean_parser/templates/go_server.jinja2

```diff
@@ -0,0 +1,263 @@
+// interface to send the log messages, accepting a call passing in the eventType and the message.
+// TODO: confirm what calls to cloud logging would be required to see if the log method signature makes sense
+// TODO: confirm if this should depend on https://github.com/mozilla-services/go-mozlog
+type Logger interface {
+	log(eventType string, message string)
+}
+
+type GleanEventsLogger struct {
+	AppID             string
+	AppDisplayVersion string
+	AppChannel        string
+	Logger            Logger
+}
+
+// exported type for public method parameters
+type RequestInfo struct {
+	UserAgent string
+	IpAddress string
+}
+
+// TODO: verify with data team what defaults make sense for user agent / ip address if they are not supplied
+var defaultRequestInfo = RequestInfo{
+	UserAgent: "Unknown",
+	IpAddress: "Unknown",
+}
+
+// structs to construct the glean ping
+type clientInfo struct {
+	TelemetrySDKBuild string `json:"telemetry_sdk_build"`
+	FirstRunDate      string `json:"first_run_date"`
+	OS                string `json:"os"`
+	OSVersion         string `json:"os_version"`
+	Architecture      string `json:"architecture"`
+	AppBuild          string `json:"app_build"`
+	AppDisplayVersion string `json:"app_display_version"`
+	AppChannel        string `json:"app_channel"`
+}
+
+type pingInfo struct {
+	Seq       int    `json:"seq"`
+	StartTime string `json:"start_time"`
+	EndTime   string `json:"end_time"`
+}
+
+type ping struct {
+	DocumentNamespace string `json:"document_namespace"`
+	DocumentType      string `json:"document_type"`
+	DocumentVersion   string `json:"document_version"`
+	DocumentID        string `json:"document_id"`
+	UserAgent         string `json:"user_agent"`
+	IpAddress         string `json:"ip_address"`
+	Payload           string `json:"payload"`
+}
+
+type metrics map[string]map[string]interface{}
+
+type pingPayload struct {
+	ClientInfo clientInfo   `json:"client_info"`
+	PingInfo   pingInfo     `json:"ping_info"`
+	Metrics    metrics      `json:"metrics"`
+	Events     []gleanEvent `json:"events"`
+}
+
+type gleanEvent struct {
+	Category  string     `json:"category"`
+	Name      string     `json:"name"`
+	Timestamp int64      `json:"timestamp"`
+	Extra     [][]string `json:"extra"`
+}
+
+func (g GleanEventsLogger) createClientInfo() clientInfo {
+	// Fields with default values are required in the Glean schema, but not used in server context
+	return clientInfo{
+		TelemetrySDKBuild: "{glean_parser_version}",
+		FirstRunDate:      "Unknown",
+		OS:                "Unknown",
+		OSVersion:         "Unknown",
+		Architecture:      "Unknown",
+		AppBuild:          "Unknown",
+		AppDisplayVersion: g.AppDisplayVersion,
+		AppChannel:        g.AppChannel,
+	}
+}
+
+func createPingInfo() pingInfo {
+	// TODO: confirm expected date format in logs,
+	// js uses Date().toISOString(), ex. "2023-12-13T20:21:55.390Z"
+	// ruby uses Time.now.utc, ex. "2023-12-13 20:21:02 UTC"
+	var now = time.Now().UTC().Format("2006-01-02 15:04:05 MST")
```

Comment:
Is there an expected date format string in the data pipeline, or will either of these work?

Reply:
This should be ISO 8601 -formatted string.
The Ruby case you're mentioning is interesting, I see it properly formatted in BQ (e.g. `2023-12-18T06:40:55.143Z`).